en/mavgen_c/message_signing_c.md

```diff
@@ -0,0 +1,168 @@
+The function offers a way for the implementation to allow unsigned packets to be accepted (and incorrectly signed packets, which might be accepted under some circumstances).
+
+The rules for what unsigned packets should be accepted is implementation specific, but it is suggested the following rules be considered:
+
+* have a mechanism for marking a particular communication channel as being secure (such as a USB connection) to allow for signing setup.
+* always accept `RADIO_STATUS` packets for feedback from 3DR radios (which don't do signing)
+
+For example:
+
+```c
+static const uint32_t unsigned_messages[] = {
+	MAVLINK_MSG_ID_RADIO_STATUS
+};
+
+static bool accept_unsigned_callback(const mavlink_status_t *status, uint32_t message_id)
+{
+	// Make the assumption that channel 0 is USB and should always be accessible
+	if (status == mavlink_get_channel_status(MAVLINK_COMM_0)) {
+		return true;
+	}
+
+	for (unsigned i = 0; i < sizeof(unsigned_messages) / sizeof(unsigned_messages[0]); i++) {
+		if (unsigned_messages[i] == message_id) {
+			return true;
+		}
+	}
+
+	return false;
+}
+```
+
+## Handling Link IDs {#handling_link_ids}
+
+The purpose of the `link_id` field in the *MAVLink 2* signing structure is to prevent cross-channel replay attacks. 
+Without the `link_id` an attacker could record a packet (such as a disarm request) on one channel, then play it back on a different channel.
+
+The intention with the link IDs is that each channel of communication between an autopilot and a GCS uses a different link ID. 
+There is no requirement that the same link ID be used in both directions however.
+
+For C implementations the obvious mechanism is to use the MAVLink channel number as the link ID. 
+That works well for an autopilot, but runs into an issue for a GCS implementation. 
+The issue is that a user may launch multiple GCS instances talking to the same autopilot via different communication links (such as two radios, or USB and a radio). 
+These multiple GCS instances will not be aware of each other, and so may choose the same link ID. 
+If that happens then a large number of correctly signed packets will be rejected by the autopilot as they will have timestamps that are older than the timestamp received for the same stream tuple on the other communication link.
+
+The solution adopted for *MAVProxy* is shown below:
+
+```
+if (msg.get_signed() and
+	self.mav.signing.link_id == 0 and
+	msg.get_link_id() != 0 and
+	self.target_system == msg.get_srcSystem() and
+	self.target_component == msg.get_srcComponent()):
+	# change to link_id from incoming packet
+	self.mav.signing.link_id = msg.get_link_id()
+```
+
+What that says is that if the current link ID in use by MAVProxy is zero, and it receives a correctly signed packet with a non-zero link ID then it switches link ID to the one from the incoming packet.
```

Comment:
To confirm, this works because each of the GCS instances is connected via a different autopilot channel, and each of those channels will have a different link id. So slaving to the autopilot id will always generate unique ids for each instance. For the case where the GCS were connected on the same channel, there wouldn't be a problem because timestamps would be correct ... and effectively the link ids would be the same.